Docker migration from tum-gis

BWibo opened the pull request:
This PR is for moving the development from https://github.com/tum-gis/3dcitydb-docker-postgis to this repo and make adaptions for future integration for automated testing and CI.

* Build image variants for 3DCityDB version v4.1 and newer and provide them to the community on e.g. DockerHub. I reserved `3dcitydb` DockerHub for us. I propose `3dcitydb/3dcitydb-pg` or `3dcitydb/3dcitydb-postgis` as image name.
* Build the images from this repo using `COPY`.
* Image variants: `debian` and `alpine`
* Use the official [Docker PostGIS](https://github.com/postgis/docker-postgis) as base image.

### ToDos
- [x] Dockerfile & scripts PostgreSQL/PostGIS
- [x] Dockerfile & scripts Oracle
- [x] Automated build and distribution using Github Actions: edge image
- [x] Automated build and distribution using Github Actions: all available postgis-docker image tags

### Whats new?

* If no `SRID` env var is passed to `docker run`, no database is created.
* `CITYDBNAME` env var was removed. Instead, the default behavior for of the postgres docker is used:
  > Unless -e POSTGRES_DB is passed to the container at startup time, this database will be named after the admin user (either   postgres or the user specified with -e POSTGRES_USER).
* I adaped the behavior for `GMLSRSNAME` from the 3DCityDB shell scripts.
  - If only `SRID` is passed to docker run `GMLSRSNAME = "urn:ogc:def:crs:EPSG::$SRID"`
  - If `SRID` and `HEIGHT_EPSG` is passed to docker run:  `GMLSRSNAME="urn:ogc:def:crs,crs:EPSG::$SRID,crs:EPSG::$HEIGHT_EPSG"`
  - If `GMLSRSNAME` is passed, the value that is passed is used.
* `POSTGIS_SFCGAL` var can be used to enable PostGIS SFCGAL (only debian images).

### Quickstart
```bash
docker run -i -t -p 5432:5432 --name cdb \
  -e POSTGRES_DB=citydb \          # Optional: Set name for the DB. IF not set, DB is named like POSTGRES_USER (default=postgres)
  -e POSTGRES_PASSWORD=postgres \  # Required: Set DB Password
  -e SRID=25832 \                  # If SRID is not set, no 3DCityDB instance is set up in the container
  -e HEIGHT_EPSG=7837 \            # Optional: Height EPSG for auto GMLSRSNAME generation
  -e GMLSRSNAME=EPSG:25832 \       # Optional: Overwrites auto generated GMLSRSNAME from SRID and HEIGHT_EPSG
  -e POSTGIS_SFCGAL=true \         # Optional: Enable SFCGAL support, only currently available in Debian images, default = false
3dcitydb/3dcitydb-pg


# Behavior GMLSRSNAME

# only SRID set              GMLSRSNAME="urn:ogc:def:crs:EPSG::$SRID"
# SRID and HEIGHT_EPSG set   GMLSRSNAME="urn:ogc:def:crs,crs:EPSG::$SRID,crs:EPSG::$HEIGHT_EPSG"
# GMLSRSNAME set             GMLSRSNAME=GMLSRSNAME
```

#### Oracle Version

The Oracle version uses the the official Oracle Database server docker image 19c (19.3.0.0) from the [Oracle Container Registry](https://container-registry.oracle.com/) as base image. Before using the docker image, you have to authenticate with your Oracle account to access to the Oracle Registry Server. A detailed instructions can be found on this [website](https://docs.oracle.com/cd/E37670_01/E75728/html/oracle-registry-server.html).

```bash
docker run -i -t -p 1521:1521 --name cdb \
  -e DB_USER=citydb \              # Optional: Set name for the oracle DB user. IF not set (default=citydb)
  -e ORACLE_PWD=citydb \           # Required: Set DB Password for the DB_USER as well as for SYS, SYSTEM and PDBADMIN
  -e SRID=25832 \                  # If SRID is not set, no 3DCityDB instance is set up in the container
  -e HEIGHT_EPSG=7837 \            # Optional: Height EPSG for auto GMLSRSNAME generation
  -e GMLSRSNAME=EPSG:25832 \       # Optional: Overwrites auto generated GMLSRSNAME from SRID and HEIGHT_EPSG
3dcitydb/3dcitydb-oracle

# The behavior of GMLSRSNAME is the same as for the PostgreSQL version. 
```

The 3DCityDB instance created above can be accessed using SQL*Plus: `sqlplus citydb/citydb@localhost:1521/ORCLPDB1`

BWibo commented:
@yaozhihang From my side everything is done besides the Github action for building images for all postgis docker images. I think your good to go adding the oracle image version.

yaozhihang commented:
> @yaozhihang From my side everything is done besides the Github action for building images for all postgis docker images. I think your good to go adding the oracle image version.

thanks @BWibo I'll add the oracle version soon.

yaozhihang commented:
While running the `docker run...`, i got the following error message. It seems to be caused by the Windows line endings in the `3dcitydb-initdb.sh`. Maybe the file should be formatted to Unix line endings 
file
![image](https://user-images.githubusercontent.com/5942971/107377041-d302ce00-6aea-11eb-92c1-ecc672042f6f.png)

yaozhihang commented:
Could we simply copy the `postgresql/DockerScripts/3dcitydb-initdb` into the `docker-entrypoint-initdb.d` for building the docker image? @BWibo Just like what you've implemented here (https://github.com/tum-gis/3dcitydb-docker-postgis/blob/master/v4.0.3/Dockerfile)
I think this will make the Dockerfile more or less simpler. Or are there any reasons why you abandon this approach?

BWibo commented:
I assue the line ending issue was cause by git checking the file out with windows ending on your system. I changed file extension to `.sh`, now this should be handled correctly by `.gitattriburtes` file.

> Just like what you've implemented here (https://github.com/tum-gis/3dcitydb-docker-postgis/blob/master/v4.0.3/Dockerfile)
I think this will make the Dockerfile more or less simpler. Or are there any reasons why you abandon this approach?

I change that as well. Your right, that was some unneeded complexity added here.
I did that when playing around with the `addictydb` etc scripts, but it is not needed anymore, as we abandoned them.

BWibo commented:
@yaozhihang @clausnagel After adding the scrects the GitHub Action for the build and push of the `edge` image on every commit to `master` seems to work now. Currently, pushed to pull request trigger a build and push as well. I'll disabled the pull requests, as this should only be done when new commits are integrated into master.

We could however think of buildilng images for pull request too and push them to another tag:
```3dcitydb/3dcitydb-pg:PULLREQUEST```

You can find the latest image here:
https://hub.docker.com/repository/docker/3dcitydb/3dcitydb-pg
```
docker pull 3dcitydb/3dcitydb-pg:edge
docker pull 3dcitydb/3dcitydb-pg:edge-alpine
```

yaozhihang commented:
@BWibo thanks for the updates. Looks very good. 

BTW, You may have seen that I've pushed the oracle version. Would you like review it? It would also be nice if you include it in the `docker-build-push-edge.yml` file as well 😊

BWibo commented:
I added a check Github Action for the Oracle image as well. I moved it to a separete file, as the process is very different from the psql image. Should I add a build and push action as well?
I am not sure if this would cause licensing issues.
If it is fine, I will setup a push to ``3dcitydb/3dcitydb-oracle:edge`` as well.

yaozhihang commented:
@BWibo Thanks very much.

The base image of the oracle version is from here https://github.com/konnecteam/docker-oracle-12c. This repo is licensed under Apache 2.0 and hence should be fine with our 3DCityDB project. I see no licensing problems to push to `3dcitydb/3dcitydb-oracle:edge`

I added the `BASE_IMAGE_TAG` to the oracle Dockerfile 7353633943a8ea4c0096f7262faa3d1ac999cd6e. Yes, keeping consistency is always not bad.

BWibo commented:
Added push for Oracle image, but I'm still not sure if this is in line with Oracle licensing. 
The repo mentioned above is not a problem, but the Oracle 12c version added to the image (see [here](https://github.com/konnecteam/docker-oracle-12c/blob/9ace5c0a298ab7108b6833f9518e66b7ad18781d/oracle_base/Dockerfile#L15)) might be.
When downloading the ``tar.gz`` from Oracle, you have to login and agree to Oracle's licensing conditions.
I'm not a licensing expert, but we should check if we are allowed to re-use the DB within our images and provide them publicly.

BWibo commented:
I added a workflow that is going to build and push image for all existing `postgis/postgis` images.
the versions are name: `3dcitydb/3dcitydb-pg:POSTGIS_IMAGE-TAG-3DCITYDB_RELEASE_VERSION`, for instance `3dcitydb/3dcitydb-pg:13-3.1-4.0.3` or `3dcitydb/3dcitydb-pg:9.6-2.51-4.0.3` and so on.
The workflow is triggered on every release of the 3DcityDB on Github.

BWibo commented:
From my side this is ready for review. There is just the licensing issue rergarding the Oracle images (@yaozhihang ).

yaozhihang commented:
@BWibo one small question: is this file "postgresql/build.sh" necessary? I would like to remove it, because it is platform(OS)-dependent script. If you need an easy way to run the docker build, maybe we could add two tasks e.g. `buildOracleCitydbDockerImage` and  `buildPosgisCitydbDockerImage` in the `build.gradle` file. There is a nice [gradle plugin](https://bmuschko.github.io/gradle-docker-plugin/) for this purpose.

BWibo commented:
No, it's not required. Actually, it should not be in the repo at all, it's just a local convenience script. I removed and ignored it.

yaozhihang commented on postgresql/DockerScripts/3dcitydb-initdb.sh:
```diff
@@ -0,0 +1,120 @@
+    echo
+    echo "!!! WARNING: POSTGIS_SFCGAL was set, but SFCGAL is not available on Alpine Linux image variants."
+    echo "!!! Please use the Debian based image variants if you require SFCGAL support."
+    SFCGAL=true
```
@BWibo thanks. I have only one question after my final review.  From my understanding, should this line be `SFCGAL=false`?

BWibo replied:
Yes, thx for spotting this, I fixed it.
